src/library/coverart.h

```diff
@@ -30,48 +31,80 @@ struct CoverInfo {
 
     static const quint16 kNullImageHash;
 
-    CoverInfo() : source(UNKNOWN),
-                  type(NONE),
-                  coverLocation(QString()),
-                  trackLocation(QString()),
-                  // This default value is fine: qChecksum(NULL, 0) is 0.
-                  hash(CoverInfo::kNullImageHash) {}
+    CoverInfoRelative()
+            : source(UNKNOWN),
+              type(NONE),
+              // This default value is fine: qChecksum(NULL, 0) is 0.
+              hash(kNullImageHash) {
+    }
 
-    bool operator==(const CoverInfo& other) const {
+    bool operator==(const CoverInfoRelative& other) const {
         return other.source == source &&
                 other.type == type &&
                 other.coverLocation == coverLocation &&
-                other.trackLocation == trackLocation &&
                 other.hash == hash;
     }
-    bool operator!=(const CoverInfo& other) const {
+    bool operator!=(const CoverInfoRelative& other) const {
         return !(*this == other);
     }
 
     Source source;
     Type type;
-    QString coverLocation; // Relative path, starting from trackLocation
-    QString trackLocation;
+    QString coverLocation; // relative path, from track location
     quint16 hash;
 };
 
-struct CoverArt {
-    CoverArt() {}
+class CoverInfo : public CoverInfoRelative {
+  public:
+    CoverInfo() {}
+
+    CoverInfo(const CoverInfoRelative& base, const QString& tl)
```

Comment:
Yes, you are right, it is composing data. 

I have squeezed this forth and back. The problem in this case with a composition is that three classes are involved. The member access will be either cluttered like this: 
coverArt.CoberInfo.CoverInforelative.hash or we would need a bunch of redundant getter and setter functions, which hide the composition nature anyway. 
A composition becomes handy, if the composed member is a base class of a set of possible classes, Since this is not the case here, I like to keep the Inheritance for now. 
I have added the missing virtual destructors.

Does this work for you? If not, I will go the extra mile and introduce the getter and setter for the composing model.